00_Base/src/mapper/BaseTransactionMapper.ts

```diff
@@ -0,0 +1,141 @@
+import { SequelizeTariffRepository, Tariff, Transaction, TransactionEvent } from '@citrineos/data';
+import { TransactionEventRequest } from '@citrineos/base';
+import { TokenDTO } from '../model/DTO/TokenDTO';
+import { OcpiLocationRepository } from '../repository/OcpiLocationRepository';
+import { TokensRepository } from '../repository/TokensRepository';
+import { ILogObj, Logger } from 'tslog';
+import { Price } from '../model/Price';
+import { Session } from '../model/Session';
+import { Tariff as OcpiTariff } from '../model/Tariff';
+import { TariffKey } from '../model/OcpiTariff';
+import { TariffsService } from '../services/tariffs.service';
+import { LocationDTO } from '../model/DTO/LocationDTO';
+import { LocationsService } from '../services/locations.service';
+
+export abstract class BaseTransactionMapper {
+  protected constructor(
+    protected logger: Logger<ILogObj>,
+    protected locationsService: LocationsService,
+    protected ocpiLocationsRepository: OcpiLocationRepository,
+    protected tokensRepository: TokensRepository,
+    protected tariffRepository: SequelizeTariffRepository,
+    protected tariffsService: TariffsService,
+  ) {
+  }
+
+  protected async getLocationDTOsForTransactions(
+    transactions: Transaction[],
+  ): Promise<Map<string, LocationDTO>> {
+    const transactionIdToLocationMap: Map<string, LocationDTO> = new Map();
+    for (const transaction of transactions) {
+      const chargingStation = await transaction.$get('station');
+      if (!chargingStation || !chargingStation.locationId) {
+        continue;
+      }
+
+      const location = await this.locationsService.getLocationById(chargingStation.locationId);
+
+      if (!location || !location.data) {
+        continue;
+      }
+
+      transactionIdToLocationMap.set(transaction.transactionId, location.data);
+    }
+    return transactionIdToLocationMap;
+  }
+
+  protected async getTokensForTransactions(
+    transactions: Transaction[]
+  ): Promise<{ [key: string]: TokenDTO }> {
+    const transactionIdToTokenMap: { [transactionId: string]: TokenDTO } = {};
+
+    for (const transaction of transactions) {
+      const startEvent = transaction.transactionEvents?.find(
+        (event) => event.eventType === 'Started'
+      );
+      if (startEvent?.idToken) {
+        const tokenDto = await this.tokensRepository.getTokenDtoByIdToken(
+          startEvent.idToken.idToken,
+          startEvent.idToken.type
+        );
+        if (tokenDto) {
+          transactionIdToTokenMap[transaction.transactionId] = tokenDto;
+        }
+      }
+    }
+
+    return transactionIdToTokenMap;
+  }
+
+  protected async getTariffsForTransactions(
+    transactions: Transaction[]
+  ): Promise<Map<string, Tariff>> {
+    const transactionIdToTariffMap = new Map<string, Tariff>();
+    const uniqueStationIds = [...new Set(transactions.map(t => t.stationId))];
+
+    const stationIdToTariffMap = new Map<string, Tariff>();
+    await Promise.all(
```

Comment:
Why is the `Promise.all` needed here? Is the idea that each fetch is executed concurrently?

Otherwise, couldn't we have just done something like:
```
    for (const stationId of uniqueStationIds) {
      const tariff = await this.tariffRepository.findByStationId(stationId);
      if (tariff) {
        stationIdToTariffMap.set(stationId, tariff);
      }
    }
```

Reply:
Good point, I actually did Promise.all thinking that it would run in parellel but I forgot that Javascript doesn't do that 😅 

I looked into it a bit, and it seems that Promise.all is still better to use when executing multiple promises, just due to the [way it handles the process](https://stackoverflow.com/a/54291660) (fails earlier, supports error handling better). Doesn't quite apply to this situation but learning new things every day. Adjusted the logic to do fetch ALL by the way as you pointed out, so we should be good on that regard for this PR!

Reply:
You can run the requests in parallel using Promise.all - but you cannot write it the way because currently your promises are being passed into the `.map()` function. What you need is `Promise.all(PROMISE_LIST_HERE)`... so you just need to make sure that you do a loop and put together your `const promiseList: Promise<?> = []` and then when you execute promise all it should open a separate node process for each request

Reply:
I think it is limited to the number of cores you have available in the end

Reply:
Oh you can! That's cool, didn't know. Is that a runtime thing? I've been reading online that you can't run parallel for node, am I just reading outdate information?

Reply:
Promise.all should trigger all requests in a non blocking fashion and wait for all requests to complete before returning, so technically they will be running in parallel. I think you can also start a new node process if you really wanted it to run on another thread on your processor though which can also be done by doing like `exec('npm run something-else`)` but I think Promise.all should suffice in triggering the requests in parallel